pandas/src/algos_take_helper.pxi

```diff
@@ -4947,3 +4947,22 @@ def take_2d_multi_object_object(ndarray[object, ndim=2] values,
                     out[i, j] = fv
                 else:
                     out[i, j] = values[idx, idx1[j]]
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def take_bounds_check(int64_t[:] indexer, int64_t[:] out, int64_t n):
+    cdef:
```

Comment:
I think the right way to do this is to make a second set of `take_dtype_dtype` functions with bounds checking, then it's all in one pass and avoids any extra allocations.  Adds a lot of new cython, but shouldn't be too painful with the templating.